### PR TITLE
`match_same_arms` fix

### LINKED ISSUE #860: False positive on match_same_arms
Example: Any kind of `xor`-like handling, e.g. let `Foo` be one variant of an enum:

```
match (x, y) {
    (Foo(_), Foo(_)) => None,
    (Foo(x), _) |
    (_, Foo(x)) => Some(x),
    _ => None
}
```

This will trigger the lint, because the _position_ of the arms is not taken into account.

@regexident suggested to only lint adjacent same arms. Another option would be to check if those arms are disjunct and can be reordered without changing semantics, but that is obviously much much harder to do in the general case.

### LINKED ISSUE #1140: match_same_arms: suppress when reordering arms?
`match_same_arms` will not consider the order of match arms.

Since it is specified that arms are checked in order, this leads to suggestions with two potential problems:
- less optimal performance, when expensive arm conditions are checked too early
- changes in semantics

Example for the latter:

```
fn main() {
    let x = match 2 {
        1 ... 3 => false,
        0 ... 5 => true,
        _ => false
    };
}
```

If this seems bad anyway because of overlap, consider the slightly more complex

```
fn main() {
    let y = 1;
    let x = match 2 {
        1 ... 3 => false,
        0 ... 5 if y == 2 => true,
        _ => false
    };
}
```

### PR DESCRIPTION
**Pull request by Jarcho**
fixes #860
fixes #1140

changelog: Don't lint `match_same_arms` when an interposing arm's pattern would overlap

### CODE REVIEW

**Comment by camsteffen** [comment]
Could we use rust-lang/rust#89570 when that is ready?

**Comment by Jarcho** [comment]
A quick look says no. Two things that are different

1. Exhaustiveness checking is comparing patterns checking if there is anything which doesn't overlap. This is checking if there is anything which does overlap.
2. Exhaustiveness checking does a rolling comparison (p1 with p2, p1+p2 with p3...). This needs a full pairwise comparison to work.

**Comment by Jarcho** [comment]
Basically done now. Just need to clarify the message for which direction the patterns need to be merged. In some cases it can only be merged in one direction. e.g.

```rust
match foo {
    Foo::A(0) => x, // can be Foo::A(0) | Foo::B(0)
    Foo::A(_) => y,
    Foo::B(0) => x, // can't merge Foo::A(0) here
}
```

If `Foo::A(_)` were changed to `Foo::B(_)` the merge would have to go in the opposite direction. If it were instead `Foo::C` then either change would work. I'm not really sure the best way to display that in the message.

**Comment by xFrednet** [comment]
Hey @Jarcho, this PR is on my todo list but sadly a bit too big to review on the side (That's why I haven't left a review yet). I'll be free in about two weeks, would it be alright if I keep this in my backlog until then? Otherwise, I can also see if someone else wants to review this :upside_down_face:

**Comment by Jarcho** [comment]
The issue's been open for more than five years. I think it can wait a couple weeks. Now get back to your thesis.

**Comment by xFrednet** [comment]
:joy: Will do, thanks! :upside_down_face:

**Review by xFrednet — COMMENTED** [review]
Alright, I've looked through the implementation and everything looks good to me. I've marked some smaller nits but nothing major. I still have to go through the tests and test output. But you can rebase if you want.

Thank you very much for giving me the time to write on my thesis. It took quite some work, But I'm happy with the result.

File: clippy_lints/src/matches.rs [hunk 1]
```diff
@@ -2242,6 +2242,203 @@ fn test_overlapping() {
...
+            && match other.bounds {
+                RangeEnd::Included => self.start <= other.end,
+                RangeEnd::Excluded => self.start < other.end,
+            }
+    }
+
+    fn is_empty(&self) -> bool {
+        match self.bounds {
+            RangeEnd::Included => false,
+            RangeEnd::Excluded => self.start == self.end,
```
[the remark this answers is not in the document]
Comment (Jarcho) [reply clippy_lints/src/matches.rs:2292]: So it does. I never checked what happened in this case.

File: clippy_lints/src/matches.rs [hunk 2]
```diff
@@ -2242,6 +2242,203 @@ fn test_overlapping() {
...
+            },
+            PatKind::Lit(e) => match &e.kind {
+                ExprKind::Lit(lit) => match lit.node {
+                    LitKind::Str(sym, _) => Self::LitStr(sym),
+                    LitKind::ByteStr(ref bytes) => Self::LitBytes(&**bytes),
+                    LitKind::Byte(val) => Self::LitInt(val.into()),
+                    LitKind::Char(val) => Self::LitInt(val.into()),
+                    LitKind::Int(val, _) => Self::LitInt(val),
+                    LitKind::Bool(val) => Self::LitBool(val),
+                    LitKind::Float(..) | LitKind::Err(_) => Self::Wild,
```
[the remark this answers is not in the document]
Comment (Jarcho) [reply clippy_lints/src/matches.rs:2366]: Worst case is a false negative here. I really doubt it's worth the effort to track which values are used.

File: clippy_lints/src/matches.rs [hunk 3]
```diff
@@ -2457,6 +2456,25 @@ impl<'hir, 'arena> ResolvedPat<'hir, 'arena> {
             (Self::LitBool(x), Self::LitBool(y)) => x == y,
             (Self::Range(ref x), Self::Range(ref y)) => x.overlaps(y),
             (Self::Range(ref range), Self::LitInt(x)) | (Self::LitInt(x), Self::Range(ref range)) => range.contains(x),
+            (Self::Slice(lpats, None), Self::Slice(rpats, None)) => {
+                lpats.len() == rpats.len() && lpats.iter().zip(rpats.iter()).all(|(x, y)| x.can_also_match(y))
+            },
+            (Self::Slice(pats, None), Self::Slice(front, Some(back)))
+            | (Self::Slice(front, Some(back)), Self::Slice(pats, None)) => {
+                if pats.len() < front.len() + back.len() {
```
[the remark this answers is not in the document]
Comment (Jarcho) [reply clippy_lints/src/matches.rs:2464]: At this point I don't know if I took that into account. I'll figure that out.

File: tests/ui/match_same_arms2.rs [hunk 4]
```diff
@@ -166,4 +166,17 @@ fn match_expr_like_matches_macro_priority() {
     };
 }
 
-fn main() {}
+fn main() {
```
[the remark this answers is not in the document]
Comment (Jarcho) [reply tests/ui/match_same_arms2.rs:169]: Constants can change, so it's probably better to not deal with them.

File: clippy_lints/src/matches.rs (hunk 2 above)
[the remark this answers is not in the document]
Comment (xFrednet) [reply clippy_lints/src/matches.rs:2366]: Agreed, I just found it funny :)

**Review by xFrednet — COMMENTED** [review]
Okay, I've also looked through the test and the output looks really clean :+1:. This should be ready for merging after a rebase and after the comments have been addressed :upside_down_face:

File: tests/ui/match_same_arms2.rs (hunk 4 above)
[the remark this answers is not in the document]
Comment (xFrednet) [reply tests/ui/match_same_arms2.rs:169]: Agreed, then I would just like to see a test with negative values :)

**Comment by xFrednet** [comment]
Hey @Jarcho, this is a ping from triage. Do you plan to continue working on this PR? The direction and changes still look good to me, besides the NITs :upside_down_face:

**Comment by Jarcho** [comment]
Still need to look into your last point. It's on the list of things to do.

File: clippy_lints/src/matches.rs (hunk 3 above)
[the remark this answers is not in the document]
Comment (Jarcho) [reply clippy_lints/src/matches.rs]: It should be `<`. Added a comment about that.

File: tests/ui/match_same_arms2.rs (hunk 4 above)
[the remark this answers is not in the document]
Comment (Jarcho) [reply tests/ui/match_same_arms2.rs]: Done. Negative values are currently treated as a wild match. Something to fill in later, but I see no reason to block this on them.

**Review by xFrednet — APPROVED** [review]

**Comment by xFrednet** [comment]
Looks good to me! Thank you for the refactoring. I've also tested this implementation with lintcheck and everything looks good. :+1: :muscle: 

@bors r+

**Pull request merged**